### ISSUE TITLE
Patch 4.0

### ISSUE THREAD

**Issue by Jonathonathon** [zininzinin/AutoTrimps#207]
The new patch is going to have some pretty big changes, including masteries and late game changes.  I wanted to open this issue in preparation for when the patch drops:

Current Patch Notes:  http://trimpstest.bitballoon.com/updates.html
Reddit Current Patch Thread:  https://www.reddit.com/r/Trimps/comments/5dppce/patch_40_test_server_3/

Currently the biggest issue I've seen is that the changes to Foreman 2 completely break AutoTrimps ability to buy buildings.  This might be a little premature, but with the really big changes coming in the next patch I wanted to try and get out in front of it.

**Comment by genbtc** [comment]
The variable has been changed from foreman2 to foreman, ahead of time, in the genBTC branch: https://github.com/zininzinin/AutoTrimps/issues/208
This means that before the patch, it will not Auto click the "Manual-Build" button when you have Foremany1 bought now.  Which should probably be OK, since when I made it, I was targeting Foremany2 to be conservative about making the script not-click the build button.

**Comment by Jonathonathon** [comment]
Awesome, appreciate the early fix.  After the patch is released I'm hoping to go through it and update this ticket with a bit more info, hopefully that'll help.

**Comment by Jonathonathon** [comment]
Just a few quick notes for the new patch, I'm sure you're already working on all of this but I'm hoping for a dialogue so other AutoTrimps users can keep track of incoming/recent changes.

**New Job:  Magamancer**
I actually don't see this one getting that much use, as when it would become relevant in a zone auto portal will undoubtedly have kicked in by then.  I'd be interested to hear other schools of thought on this though.

**New Building: Dimensional Generator**
After unlocking Hybrid it pretty much runs itself.  Could we maybe automate spent magmite by having it purchase upgrades as they become available?  Prioritizing One and Done upgrades over Multi Upgrades would probably be idea.

Alerts can be turned off so the new alert shouldn't be that much of an issue.  I think that'll be enough to keep most of us going, hopefully these are some pretty easy changes?

Other script stuff not related to new content:

**Scryer Max Zone**
I've used the max Scryer zone option on the Scryer tab and I'm still seeing Scryer stance get used in higher levels.  Is this intentional?

~~**Minimum Zone Portal Option**~~
~~Can you bring back the minimum warp zone function?  Now that it's gone, my He/Hr portals immediately after I've ended a challenge (running Decay right now) at level 56 instead of the upper 300s where I should be.  I'd rather have a minimum than a % buffer as it's slightly more efficient.~~
Scratch this, think I just had a weird visual issue going on.  I'm seeing it now.

**Comment by Jonathonathon** [comment]
Another thing I noticed, since everyone is doing Nurseries all at once now, could AutoTrimps maybe add them two at a time if Double Build is unlocked?  Should allow us to construct them twice as fast.

**Comment by genbtc** [comment]
Ill do that first cause its easy. there was a plan to construct like 500 or however many at once but its hard to code. Thanks for outlining the stuff, keep it coming, I do them in easyness-order.  There is a lot to be desired with the AutoStance + scryer now (B vs. H vs. S vs. D) and something needs to change, i'll take a look at the scryer max zone next.
Some tips for anyone else reading: 
Ive been using "No Nurseries Until: z340" and cap "Max nurseries" to 500 with good success. (I'm @ 13.8T HE) but the +100 coords (pegs the gather button on science) is enough to carry my 30 Overkill all the way to at least z340ish with 0 nurseries.
Also using "Warpstation cap" becomes a much better idea since gigas were removed after z230, and way trimps come from the dimensional generator.

**Comment by Jonathonathon** [comment]
Can second the warpstation cap, especially while you're figuring out your giga/warp strategy after the patch.  Out of curiosity, what distribution are you using for Perks after the patch?  Also for what it's worth I think a magmite spender is probably a higher priority than stance work since a percent of it is lost after every portal.

**Comment by genbtc** [comment]
Good point on magmite. I havent decided on perks but i'm using Truth (late) and it seems fine. Also, Worker Ratio strategy needs to be changed I think. I've reverted to 1/1/1 for simplicity of figuring things out for now.

**Comment by Jonathonathon** [comment]
Alright I'll give it a shot.  Thanks again for updating this script and keeping us in the loop on what you have planned for it, I know myself and a bunch of people on Reddit really appreciate all your work.

**Comment by genbtc** [comment]
Magmite Spender is in. Part 1 buys any permanent one-and-done upgrades in order from most expensive to least. Part 2 then finds/buys the cheapest non-permanent multi-upgrade and repeats itself until you cant buy anymore.

IMO you should be responsible for buying your own permanent upgrades, or theres no way to please everybody with its decisions. If you cant get enough Magma to afford the perma-ones in 1 run, its going to spend the rest on non-perms, etc.

**Comment by porkbob** [comment]
I've noticed Autohomicide (or perhaps Automaps? or perhaps some interaction between the two?) will sometimes (but not always) abandon soldiers instantly upon entering a new zone in order to go to maps for prestige or map bonus, _even if_ the next batch of troops isn't finished training yet.

In the past it has seemed like the default behavior is to continue attacking in the zone until either the troops die or the next group is ready. I haven't been able to figure out what might be causing this to no longer happen sometimes. I'll keep trying to see if I can notice any pattern, but in the meantime I figured I should point it out since you'd probably only be able to figure it out from looking under the hood anyway.

**Comment by porkbob** [comment]
Not entirely related to the new patch, but could you put in a simple display for current He/hr % out of lifetime He, so we could gauge how our builds/runs are turning out at a glance? You could just put it in the battle area like hider/anso did, unless you have a better idea of somewhere to locate it. Maybe even include current run total He % out of lifetime He, too, while you're there?

**Comment by porkbob** [comment]
Also a setting to say "after this zone buy golden battle for the rest of the run".

**Comment by porkbob** [comment]
Another thought: if we are trying to go to from zone to map for prestige/map bonus and the next batch of trimps is finished breeding, then go ahead and abandon the soldiers, instead of what is sometimes a long, slow wait for them to die naturally.

**Comment by Jonathonathon** [comment]
Hey porkbob, I like your ideas but could you post these by creating a separate issue?  I'd like to keep issues/ideas/bugs in separate places so it's a little less overwhelming and it doesn't get lost in the shuffle and I wanted this ticket to be about handling the 4.0 changes..

As to your point genbtc, I see what you're getting at but I think that if people want to buy their own permanent upgrades they should be fine with running manual for a while.  The issue here is that no other resource dwindles and is lost over time, which is why I think the automation is so important for idle players.  I especially like that the base upgrades are purchased from least to most expensive, good call there.  I'll be running this tonight to see if there are any unforeseen issues but your implementation looks spot on.

I still can't think of a good way to handle Magmamancers and I can't really imagine any idle users using them.  The unfortunate part is that you have to have these guys allocated for the entire time which sucks.  This is such a low priority I kind of think you should just move on from this until someone can suggest an elegant and simple solution.

**Comment by genbtc** [comment]
My implementation was NOT spot on, and broke Autoportal if you had any permanent upgrades still available.
Refresh, fixed as of now. I am reading the other stuff.

**Comment by porkbob** [comment]
Well, Jonathonathon, before going to bed last night genr8_ linked this thread in the discord channel and said to give him ideas to code here, so that's what I did. :P

**Comment by Jonathonathon** [comment]
Ah alright no worries then. Like I said, just didn't want anything to get lost in the shuffle.
